lib/cache/memory-cache-store.js

```diff
@@ -0,0 +1,417 @@
+      value,
+      this.#maxEntrySize
+    )
+
+    // Remove the value if there was some error
+    writable.on('error', (err) => {
+      values.filter(current => value !== current)
+      if (this.#errorCallback) {
+        this.#errorCallback(err)
+      }
+    })
+
+    writable.on('bodyOversized', () => {
+      values.filter(current => value !== current)
+    })
+
+    return writable
+  }
+
+  /**
+   * @param {string} origin
+   */
+  deleteByOrigin (origin) {
+    this.#data.delete(origin)
+  }
+
+  /**
+   * Gets all of the requests of the same origin, path, and method. Does not
+   *  take the `vary` property into account.
+   * @param {import('../../types/dispatcher.d.ts').default.RequestOptions} req
+   * @returns {MemoryStoreValue[] | undefined}
+   */
+  #getValuesForRequest (req, makeIfDoesntExist) {
+    // https://www.rfc-editor.org/rfc/rfc9111.html#section-2-3
+    let cachedPaths = this.#data.get(req.origin)
+    if (!cachedPaths) {
+      if (!makeIfDoesntExist) {
+        return undefined
+      }
+
+      cachedPaths = new Map()
+      this.#data.set(req.origin, cachedPaths)
+    }
+
+    let values = cachedPaths.get(`${req.path}:${req.method}`)
+    if (!values && makeIfDoesntExist) {
+      values = []
+      cachedPaths.set(`${req.path}:${req.method}`, values)
+    }
+
+    return values
+  }
+
+  /**
+   * Given a list of values of a certain request, this decides the best value
+   *  to respond with.
+   * @param {import('../../types/dispatcher.d.ts').default.RequestOptions} req
+   * @param {MemoryStoreValue[]} values
+   * @returns {MemoryStoreValue | undefined}
+   */
+  #findValue (req, values) {
+    /**
+     * @type {MemoryStoreValue}
+     */
+    let value
+    const now = Date.now()
+    for (let i = values.length - 1; i >= 0; i--) {
+      const current = values[i]
+      const currentCacheValue = current.opts
+      if (now >= currentCacheValue.deleteAt) {
+        // We've reached expired values, let's delete them
+        this.#entryCount -= values.length - i
+        values.length = i
+        break
+      }
+
+      let matches = true
+
+      if (currentCacheValue.vary) {
+        if (!req.headers) {
+          matches = false
+          break
+        }
+
+        for (const key in currentCacheValue.vary) {
+          if (currentCacheValue.vary[key] !== req.headers[key]) {
+            matches = false
+            break
+          }
+        }
+      }
+
+      if (matches) {
+        value = current
+        break
+      }
+    }
+
+    return value
+  }
+}
+
+class MemoryStoreReadableStream extends Readable {
+  /**
+   * @type {MemoryStoreValue}
+   */
+  #value
+  /**
+   * @type {Buffer[]}
+   */
+  #chunksToSend = []
+
+  /**
+   * @param {MemoryStoreValue} value
+   */
+  constructor (value) {
+    super()
+
+    if (value.readLock) {
+      throw new Error('can\'t read a locked value')
+    }
+
+    this.#value = value
+    this.#chunksToSend = [...this.#value.body, null]
```

Comment:
if body is null, what happens then? 


```suggestion
    this.#chunksToSend = value?.body ? [...value.body, null] : [null]
```